src/log.rs

```diff
@@ -81,13 +84,327 @@ type CompatFilterStream<T, R> = Compat01As03<FilterStream<T, R>>;
+            BlockNumber::Earliest => Some(0),
+            BlockNumber::Number(value) => Some(value.as_u64()),
+            BlockNumber::Latest | BlockNumber::Pending => None,
+        };
+        let end_block = match to_block {
+            BlockNumber::Earliest => None,
+            BlockNumber::Number(value) => Some(value.as_u64()),
+            BlockNumber::Latest | BlockNumber::Pending => {
+                let latest_block = web3.eth().block_number().compat().await?;
+                Some(latest_block.as_u64())
+            }
+        };
+
+        let next = match (start_block, end_block) {
+            (Some(page_block), Some(end_block)) => PastLogsStream::Paging(PastLogsPager {
+                web3,
+                to_block,
+                block_page_size,
+                filter,
+                page_block,
+                end_block,
+            }),
+            _ => PastLogsStream::Querying(web3, filter.build()),
+        };
+
+        Ok(next)
+    }
+}
+
+/// Internal state for paging though past logs.
+struct PastLogsPager<T: Transport> {
+    web3: Web3<T>,
+
+    /// The `to_block` specified by the log filter.
+    to_block: BlockNumber,
+    /// The block page size being used for queries.
+    block_page_size: u64,
+    /// The web3 filter used for retrieving the logs.
+    filter: FilterBuilder,
+
+    /// The block number for the next page.
+    ///
+    /// This value can be `None` if it `from_block` was specified as `Latest` or
+    /// `Pending`.
+    page_block: u64,
+    /// The last block used for pagination. This is slightly different than
+    /// `to_block` as this must be a concrete block number (and can't be block
+    /// aliases such as `Earliest` or `Latest`).
+    ///
+    /// This has a value of `None` if it has not yet been retrieved (in case
+    /// `to_block` is `Latest` or `Pending`).
+    end_block: u64,
+}
+
+impl<T: Transport> PastLogsPager<T> {
+    async fn next_page(&mut self) -> Result<Option<Vec<Log>>, ExecutionError> {
+        debug_assert!(
+            self.block_page_size != 0,
+            "pager should never be constructed with 0 block page size",
+        );
+
+        while self.page_block <= self.end_block {
+            // NOTE: Log block ranges are inclusive.
+            let page_end = self.page_block + self.block_page_size - 1;
+            let page_to_block = if page_end < self.end_block {
+                BlockNumber::Number(page_end.into())
+            } else {
+                // NOTE: The last page is handled a bit differently by using the
+                //   `to_block` that was originally specified to the builder.
+                //   This is done because the `end_block` is determined when the
+                //   pager is created, and it is possible that the `to_block`
+                //   was specified as "latest" or "pending" which may have
+                //   changed since the paging started.
+                self.to_block
```

Comment:
Can't we get rid of this block and always use `self.end_block`. That's what I meant previously that changing the behavior to determine the last block on creation would simplify this part.

```rust
let page_to_block = std::cmp::min(page_end, self.end_block);
```

Reply:
Ah, sorry, I misunderstood. The only thing I see with that is that is what does "pending" mean in that case? `latest_block + 1`? Do nodes even allow such a query?

Reply:
Your argument makes sense. We can keep it like this.